### PR TITLE
Exception handling in query execution

### PR DESCRIPTION
**Pull request by dengyuchenkit**
This PR adds exception handling during execution.
This branch is rebased on #1017. The exception handling part is in the last commit. I will keep it up-to-date with that PR and check for more cases these days. 
```
postgres=# create table foo(id int);
CREATE TABLE
postgres=# create table foo(id int);
Table foo already exists
postgres=# 

```

### CODE REVIEW

**Comment by dengyuchenkit** [comment]
Hi all,

I am having a build error with LLVM and wondering is there anyone familiar
with it?

I reinstalled LLVM 3.7 from source, then the error comes out. I tried
uninstall and reinstall the LLVM 3.7 as in the packages.sh, but the error
still exists.

The error is :
../lib/libpeloton.so.0.0.5: undefined reference to `typeinfo for
llvm::Instruction'
../lib/libpeloton.so.0.0.5: undefined reference to `typeinfo for
llvm::SectionMemoryManager'
../lib/libpeloton.so.0.0.5: undefined reference to `typeinfo for
llvm::UnaryInstruction'
../lib/libpeloton.so.0.0.5: undefined reference to `typeinfo for
llvm::CmpInst'
collect2: error: ld returned 1 exit status
src/CMakeFiles/peloton-bin.dir/build.make:149: recipe for target
'bin/peloton' failed
make[2]: *** [bin/peloton] Error 1

The platform information:
INFO Using built-in specs.
COLLECT_GCC=/usr/bin/c++
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/5/lto-wrapper
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu
5.4.0-6ubuntu1~16.04.5'
--with-bugurl=file:///usr/share/doc/gcc-5/README.Bugs
--enable-languages=c,ada,c++,java,go,d,fortran,objc,obj-c++ --prefix=/usr
--program-suffix=-5 --enable-shared --enable-linker-build-id
--libexecdir=/usr/lib --without-included-gettext --enable-threads=posix
--libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu
--enable-libstdcxx-debug --enable-libstdcxx-time=yes
--with-default-libstdcxx-abi=new --enable-gnu-unique-object
--disable-vtable-verify --enable-libmpx --enable-plugin --with-system-zlib
--disable-browser-plugin --enable-java-awt=gtk --enable-gtk-cairo
--with-java-home=/usr/lib/jvm/java-1.5.0-gcj-5-amd64/jre --enable-java-home
--with-jvm-root-dir=/usr/lib/jvm/java-1.5.0-gcj-5-amd64
--with-jvm-jar-dir=/usr/lib/jvm-exports/java-1.5.0-gcj-5-amd64
--with-arch-directory=amd64 --with-ecj-jar=/usr/share/java/eclipse-ecj.jar
--enable-objc-gc --enable-multiarch --disable-werror --with-arch-32=i686
--with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib
--with-tune=generic --enable-checking=release --build=x86_64-linux-gnu
--host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 5.4.0 20160609 (Ubuntu 5.4.0-6ubuntu1~16.04.5)

Selected compiler gcc 5.4.0
-- Boost version: 1.58.0
-- Found the following Boost libraries:
--   system
--   filesystem
--   thread
--   chrono
--   date_time
--   atomic
-- Found gflags  (include: /usr/include, library:
/usr/lib/x86_64-linux-gnu/libgflags.so)
-- Found PROTOBUF Compiler: /usr/bin/protoc
-- Found libevent pthread (include: /usr/include, library:
/usr/lib/x86_64-linux-gnu/libevent_pthreads.so)
-- Found libevent (include: /usr/include, library:
/usr/lib/x86_64-linux-gnu/libevent.so)
-- Could NOT find Doxygen (missing:  DOXYGEN_EXECUTABLE)
-- Found jemalloc (include: , library:
/usr/lib/x86_64-linux-gnu/libjemalloc.so)
-- Found LLVM 3.7.1
-- 
-- ******************* Peloton Configuration Summary *******************
-- General:
--   Version           :   0.0.5
--   Git               :   e8fa8e7
--   System            :   Linux
--   Compiler          :   /usr/bin/c++ (gcc 5.4.0)
--   Release CXX flags :   -O3 -DNDEBUG -std=c++1y
-D__PELOTONFILE__='"$(subst
/home/yuchen/CMU/DatabaseGroup/cmu_peloton/peloton/,,$(abspath $<))"'
-fdiagnostics-color=auto -fPIC -Wall -Wextra -Werror -mcx16
-Wno-invalid-offsetof -march=native -Wno-strict-aliasing
-Wno-implicit-fallthrough
--   Debug CXX flags   :   -g -O0 -ggdb -fno-omit-frame-pointer
-fno-optimize-sibling-calls -std=c++1y -D__PELOTONFILE__='"$(subst
/home/yuchen/CMU/DatabaseGroup/cmu_peloton/peloton/,,$(abspath $<))"'
-fdiagnostics-color=auto -fPIC -Wall -Wextra -Werror -mcx16
-Wno-invalid-offsetof -march=native -Wno-strict-aliasing
-Wno-implicit-fallthrough
--   Build type        :   Release
-- 
--   BUILD_docs        :   ON
-- 
-- Dependencies:
--   Linker flags      :
--   Boost             :   Yes (ver. 1.58)
--   glog              :   Yes
--   gflags            :   Yes
--   protobuf          :   Yes (ver. 2.6.1)
--   capnproto         :   No
--   LLVM              :   Yes (ver. 3.7.1)
-- 
-- Documentaion:
--   Doxygen           :   No
--   config_file       :
-- 
-- Install:
--   Install path      :
/home/yuchen/CMU/DatabaseGroup/cmu_peloton/peloton/build/install
-- 
-- Configuring done
-- Generating done
-- Build files have been written to:
/home/yuchen/CMU/DatabaseGroup/cmu_peloton/peloton/build


This happens when I tried to fix the exception handling error when the
query is executed in the codegen engine. The current problem about the PR
is that it can't catch exceptions from codegen code.

I doubt this has something to do with the LLVM_ENABLE_EH option which may
be disabled by default when we install LLVM.
*LLVM_ENABLE_EH*:BOOLBuild LLVM with exception-handling support. This is
necessary if you wish to link against LLVM libraries and make use of C++
exceptions in your own code that need to propagate through LLVM code.
Defaults to OFF.from: https://llvm.org/docs/CMake.html

I tried to alter the configuration and install the LLVM 3.7 both from
source and from APT, which brings the errors above.. Does anyone have any
idea? Any idea is appreciated, thanks!

Yuchen

**Comment by apavlo** [comment]
@tcm-marcel This is the same error that you were getting before, right?

**Comment by tcm-marcel** [comment]
My error was different (it was a runtime problem). 

`undefined reference to typeinfo for xyz` can happen when mixing `-fno-rtti` and `-frtti code` - and the exceptions might use RTTI to get the exception type. I think RTTI is disabled in the Debian/Ubuntu LLVM packages, so we can't change this without recompiling it. 

Maybe there is a way to catch the exceptions in a non-RTTI way?

File: script/testing/psql/psql_ref_out.txt [hunk 1]
```diff
@@ -1,12 +1,12 @@
 DROP 0
 CREATE TABLE
 INSERT 0 1
-INSERT 0 0
+Constraint violated
```
Comment (ChTimTsubasa) [inline script/testing/psql/psql_ref_out.txt]: Why would this throwing constraint violate?

File: src/executor/plan_executor.cpp [hunk 2]
```diff
@@ -48,7 +49,7 @@ static void CompileAndExecutePlan(
   // Prepare output buffer
   std::vector<oid_t> columns;
   plan->GetOutputColumns(columns);
-  codegen::BufferingConsumer consumer{columns, context};
+  auto consumer = new codegen::BufferingConsumer(columns, context);
```
Comment (ChTimTsubasa) [inline src/executor/plan_executor.cpp]: Why are you allocating new memory here? Can you use a smart pointer?

File: src/network/postgres_protocol_handler.cpp [hunk 3]
```diff
@@ -377,7 +377,7 @@ void PostgresProtocolHandler::ExecQueryMessageGetResult(ResultType status) {
   std::vector<FieldInfo> tuple_descriptor;
   if (status == ResultType::SUCCESS) {
     tuple_descriptor = traffic_cop_->GetStatement()->GetTupleDescriptor();
-  } else if (status == ResultType::FAILURE) { // check status
+  } else if (status == ResultType::FAILURE || status == ResultType::ABORTED) { // check status
```
Comment (ChTimTsubasa) [inline src/network/postgres_protocol_handler.cpp]: Try not to use status. It should be discarded soon.

**Comment by apavlo** [comment]
Replaced by #1144

**Pull request closed without merging**